Refactor SinglyLinkedList to use generics and add features

FreeOnur opened the pull request:
Refactored the `SinglyLinkedList` class to support generics (`<T>`), making it reusable with any data type. Converted static methods to instance methods and added new methods: `InsertAfter`, `InsertBefore`, and `PosOfElement`. Enhanced `SearchNode` to use a predicate for flexible searches. Updated `SinglyLinkedListTests` to reflect these changes, including new test cases for added functionality. Overall, improved flexibility, reusability, and object-oriented design.

FreeOnur committed:
Refactor SinglyLinkedList to use generics and add features

Refactored the `SinglyLinkedList` class to support generics (`<T>`),
making it reusable with any data type. Converted static methods
to instance methods and added new methods: `InsertAfter`,
`InsertBefore`, and `PosOfElement`. Enhanced `SearchNode` to
use a predicate for flexible searches. Updated `SinglyLinkedListTests`
to reflect these changes, including new test cases for added
functionality. Overall, improved flexibility, reusability, and
object-oriented design.

FreeOnur commented:
"Bestimmen" Sie die Laufzeit-Komplexität ihrer neuen Methoden (O(?)).
Muss vlt Fall die gesamte Liste durchlaufen bis das richtige element gefunden wird was heißt das es O(n) ist.

Im Anhang ist UML Diagramm
<img width="561" height="201" alt="SingleLinkedList_UML" src="https://github.com/user-attachments/assets/206fba49-a1fe-4177-903a-eea2bac74713" />

BurakSahin2060 commented:
Clean Code:           5/5
PR:                         5/5 (sauberer PR)
Benennungen:       5/5
Code:                     5/5 (macht das was verlangt ist, ist auch einfach zu verstehen)
Stabilität:               5/5

UML Diagramm und O-Notation passen auch!

**GESAMT: 5/5 (ohne viel zu diskutieren)**

FreeOnur committed:
fixed some bugs

FreeOnur committed:
Refactor SinglyLinkedList and tests for simplicity

Removed `InsertAtSpecificPosition` and `SearchNode` methods from `SinglyLinkedList` to simplify the class. Changed `head` field from `static` to instance-level for better encapsulation. Made minor formatting updates to `PosOfElement`.

Refactored `SinglyLinkedListTests` to reflect method removals and improve maintainability. Added a `SetUp` method for shared test data, reduced redundancy, and updated tests to use `PosOfElement` for assertions. Retained and refactored key tests for clarity and abstraction.
